src/audio/kpb.c

```diff
@@ -940,6 +941,7 @@ static void kpb_init_draining(struct comp_dev *dev, struct kpb_client *cli)
 		 * in the history buffer. All we have to do now is to calculate
 		 * read pointer from which we will start draining.
 		 */
+		flags = arch_interrupt_global_disable();
```

Comment:
Do we really want to disable all IRQs on all coes ? A local spinlock_irq() would be better.. Do we also need this locking on other components too ?

Reply:
@tlauda @jajanusz fyi

Reply:
@lgirdwood, this call arch_interrupt_global_enable/disable disables IRQs on certain core only. Spinlock inside IRQ is not an option as it will stall us forever.

Reply:
@mrajwa arch_interrupt_global_enable() may disable all IRQs on all core on non Intel arch. Btw, why does the spinlock_irq() fail and stall forever.? The spinlock should protect other updates to these values and should prevent all other IRQs on that core. i.e. it should never spin forever if properly used (that means it must also be used in any irq code that touches these pointers)

Reply:
> @mrajwa arch_interrupt_global_enable() may disable all IRQs on all core on non Intel arch. 
Currently it won't happen as this interface only support Xtensa core
> Btw, why does the spinlock_irq() fail and stall forever.?
Maybe I misunderstood your idea but if what you are saying boils down to waiting inside IRQ then two thing concerns me:
1. Since init_draining is passive level we may get stuck in that IRQ
2. Even if we do it in a clever way by postponing that IRQ job it will probably be more expensive than sigle lock before these few instructions aimed for pointer calculation.

Reply:
I still see this as a locking issue, I think you should look at all the asynchronous flows within kbp and see where they can race. You are probably missing protection around some of these calculations that can either be trigger via pipeline, voice detect, DMA, IPC or other sources.

The spinlock_irq() wont wait in the IRQ handler. If we are spinning on acquiring the lock in IRQ context then our locking in KBP is broken. Generally you should be protecting (with the spinlock) anything state that the IRQ handler will change so that code running at normal context will see the state atomically.

Reply:
> I think you should look at all the asynchronous flows within kbp and see where they can race. You are probably missing protection around some of these calculations that can either be trigger via pipeline, voice detect, DMA, IPC or other sources.

Yes this is what I have done and this PR is the result of such analysis. In any other place, including draining we are fine when the IRQ happens. But at the init draining time it is dangerous.

As for the spinlock_irq(), the very first thing it does in current implementation is exactly what I did in the first commit - disable all IRQ below level 5. Then it keeps the lock in shared memory context so other cores should not tinker with our data. The later seems to be premature now as we only utilize one core now.

Reply:
please use the spinlock. it's client level API and arch_ are system level APIs not intended for audio processing components.

Reply:
Okay I buy this argument now. I have updated, is it OK for you now?